module/zfs/spa.c

```diff
@@ -9759,6 +9759,12 @@ void
 spa_event_notify(spa_t *spa, vdev_t *vd, nvlist_t *hist_nvl, const char *name)
 {
 	spa_event_post(spa_event_create(spa, vd, hist_nvl, name));
+
+	if (strcmp(name, ESC_ZFS_SCRUB_FINISH) == 0 ||
+	    strcmp(name, ESC_ZFS_RESILVER_FINISH) == 0) {
```

Comment:
To me it seems a little nonintuitive to be detecting scrub/resilver completion by when the event posts.  It's different than how we detect `zpool clear`, which could in theory rely on the ESC_ZFS_VDEV_CLEAR event, but I think it's nicer the way you've done it in `vdev_clear()`.

I get that `zfs_ereport_clear()` is related to events so detecting it in the same layer is nice, but I think it would be reasonable to do it in `dsl_scan_done()` and `vdev_rebuild_complete_sync()`.

That said, looking at where `ESC_ZFS_RESILVER_FINISH` is generated, it seems odd to me that we are generating this event when sequential rebuild completes successfully (`vdev_rebuild_complete_sync()`) or is canceled (`vdev_rebuild_cancel_sync()`).  If a verifying resilver is canceled, this event is not generated.  So I think we should not be generating it (or clearing the error history) if a sequential resilver is canceled.

You could also argue that since a sequential resilver is always followed by a scrub, we could wait to clear the error history until the scrub completes.  But it probably doesn't matter much either way.

Reply:
good point.  I'll switch to using dsl_scan_done() and vdev_rebuild_complete_sync().